pyface/color.py

```diff
@@ -0,0 +1,260 @@
+    need a dictionary key, use an appropriate channel tuple from the
+    object.
+    """
+
+    #: A tuple holding the red, green, blue, and alpha channels.
+    rgba = AlphaChannelTuple()
+
+    #: A tuple holding the red, green, and blue channels.
+    rgb = Property(ChannelTuple(), depends_on='rgba')
+
+    #: The red channel.
+    red = Property(Channel, depends_on='rgba')
+
+    #: The green channel.
+    green = Property(Channel, depends_on='rgba')
+
+    #: The blue channel.
+    blue = Property(Channel, depends_on='rgba')
+
+    #: The alpha channel.
+    alpha = Property(Channel, depends_on='rgba')
+
+    #: A tuple holding the hue, saturation, value, and alpha channels.
+    hsva = Property(AlphaChannelTuple, depends_on='rgba')
+
+    #: A tuple holding the hue, saturation, and value channels.
+    hsv = Property(ChannelTuple, depends_on='rgb')
+
+    #: A tuple holding the hue, lightness, saturation, and alpha channels.
+    hlsa = Property(AlphaChannelTuple, depends_on='rgba')
+
+    #: A tuple holding the hue, lightness, and saturation channels.
+    hls = Property(ChannelTuple, depends_on='rgb')
+
+    @classmethod
+    def from_toolkit(cls, toolkit_color, **traits):
+        """ Create a new Color object from a toolkit color object.
+
+        Parameters
+        ----------
+        toolkit_color : toolkit object
+            A toolkit color object, such as a Qt QColor or a Wx wx.Colour.
+        **traits
+            Any additional trait values to be passed as keyword arguments.
+        """
+        from pyface.toolkit import toolkit_object
+        toolkit_color_to_rgba = toolkit_object('color:toolkit_color_to_rgba')
+        rgba = toolkit_color_to_rgba(toolkit_color)
+        return cls(rgba=rgba, **traits)
+
+    def to_toolkit(self):
+        """ Create a new toolkit color object from a Color object.
+
+        Returns
+        -------
+        toolkit_color : toolkit object
+            A toolkit color object, such as a Qt QColor or a Wx wx.Colour.
+        """
+        from pyface.toolkit import toolkit_object
+        rgba_to_toolkit_color = toolkit_object('color:rgba_to_toolkit_color')
+        return rgba_to_toolkit_color(self.rgba)
+
+    def hex(self):
+        """ Provide a hex representation of the Color object.
+
+        Note that because the hex value is restricted to 0-255 integer values
+        for each channel, the representation is not exact.
+
+        Returns
+        -------
+        hex : str
+            A hex string in standard ``#RRGGBBAA`` format that represents
+            the color.
+        """
+        values = channels_to_ints(self.rgba)
+        return "#{:02X}{:02X}{:02X}{:02X}".format(*values)
+
+    def __eq__(self, other):
+        if isinstance(other, Color):
+            return self.rgba == other.rgba
+        return NotImplemented
+
+    def __str__(self):
+        return "({:0.5}, {:0.5}, {:0.5}, {:0.5})".format(*self.rgba)
```

Comment:
What is this `__str__` used for? 

This reminds me of Python 2 behaviour where `str(float)` does not roundstrip, and it has caused issues because it was not at all obvious (e.g. marshmallow Float uses `str`, and it causes issue for people who try to use marshmallow for serializing data in JSON and deserialized back, and expecting the data to stay the same... and then be surprised when they don't).

That Python 2 has since changed in Python 3 such that `str(float)` does round trip.  I am wondering whatever that had bitten people before with `str(float)` not roundtripping might bite us here.

To be safe, could we use `{!r}` here?

Reply:
The point of `__str__` is always a human-friendly representation of an object, suitable for use in a `print` statement or, in the case of TraitsUI, the default representation in a UI.  Round-tripping is always a secondary consideration.

In particular, the motivation for this is what will appear in a TraitsUI colour editor's text field.

Given that in practice the colours are going to be converted to bytes, they key thing is that the round-trip byte -> float -> str -> float -> byte works (and it does) so that the actual _color_ doesn't change.

Actually, we can even get away with `{0:3}` here.

If there is a specialist case which requires more precision, the user can override the editor's format function to give full precision.

Reply:
Ah - I remember why 5.  The CSS styles allow you to specify with 4-digit hex values if you want.  5 places allows round-tripping all of those successfully.

Reply:
I am wondering if this is actually a view layer that gets put on the model. Perhaps this conversion could be implemented in the view instead of here?

Reply:
It could, but you want a default, and since this default also makes sense for other situations where you want a human-friendly representation, I think it makes sense to have a `__str__` method.

Otherwise why does Python have a `__str__` at all?

Reply:
For a library, it allows more room for modifications later if we are more conservative about the functionality exposed. The tools are there but we may not choose to use it, e.g. namedtuple is there but we would be exposing more capability than we need if we use namedtuple for something where we only needed named attribute access, making it hard to back out later. Anyway, it was just a thought.